Add advanced password generator features

seregaxvm opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

## Description
<!--- Describe your changes in detail -->
Added options to increase control over the character groups used for password generation.

Special characters group has been split into 6 subgroups (see screenshot below). Buttons controlling these groups and extended ASCII button were moved to fineTuneBar widget and are visible only when Fine Tune button is pressed. 



## Motivation and context
A lot of sites have very particular set of characters allowed in password. This change allows to generate a password in compliance with character restrictions simply by selecting character groups.

fineTuneBar is initially hidden so additional information will not overload the UI.

Extended ASCII button was moved inside fineTuneBar because firstly, it makes sense, secondly, so UI would not change a lot. 

## How has this been tested?

- It has been tested by hand.
- Automated tests were run and passed.
- Program was build and run with `-DWITH_ASAN=ON` before and after alterations. No new memory leaks were introduced (BTW there are some memory leaks on development brunch).

## Screenshots (if appropriate):

### Before changes
![selection_206](https://user-images.githubusercontent.com/26824900/38697862-d3ca5f34-3e9b-11e8-8bb2-b878db19d3fd.png)

### After changes
![selection_207](https://user-images.githubusercontent.com/26824900/38697898-f3fa8202-3e9b-11e8-887d-fa9d83f78c03.png)

### After changes with fine tune button pressed
![selection_208](https://user-images.githubusercontent.com/26824900/38697908-f8a49054-3e9b-11e8-8acf-c9159cecfd4f.png)


## Types of changes
<!--- What types of changes does your code introduce? -->
<!--- Please remove all lines which don't apply. -->
- ✅ New feature (non-breaking change which adds functionality)

## Checklist:
<!--- Please go over all the following points. -->
<!--- Again, remove any lines which don't apply. -->
<!--- Pull Requests that don't fulfill all [REQUIRED] requisites are likely -->
<!--- to be sent back to you for correction or will be rejected.  -->
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**

droidmonkey commented:
Oh i like this, although i think a label of "Advanced..." or "Extended ASCII..." would be more appropriate.

seregaxvm commented:
Replaced FineTune button with Advanced button on the side. When clicked, it reveals advanced configurations. User can get back to simple configurations by clicking Simple button.

Added ability to create hexadecimal passwords, proposed in #1248 by splitting characters is subgroups.

### Simple mode
![selection_209](https://user-images.githubusercontent.com/26824900/38766342-6e07e79a-3fd9-11e8-893c-2ca6bdb2e839.png)
### Advanced mode
![selection_210](https://user-images.githubusercontent.com/26824900/38766345-749ed74e-3fd9-11e8-8d5b-12935e46caef.png)

TheZ3ro commented:
I really like this new UI, I will review in the next few days

glubsy commented:
What if I only want some specific characters among `<*+!?=` and only some characters among `\_|-/` (as required by some programs / websites)?

seregaxvm commented:
If it is a frequent issue, groups should be adjusted. If it is just a couple of sites/programs, manual password edit is fine. I don't think adding more buttons would be a good idea. It would be too messy.

glubsy commented:
But what if I generate a lot of passwords, and they need to be 500+ characters? 

Wouldn't a filter be more appropriate? Like a blacklist / whitelist pair or small input fields perhaps?

droidmonkey commented:
There was discussion of including a textbox where a white/black list could be applied.

seregaxvm commented:
Do you need this in GUI? Could this just be a CLI feature?

droidmonkey commented:
Yes it absolutely must be in the GUI

seregaxvm commented:
![selection_234](https://user-images.githubusercontent.com/26824900/39785267-2db4cede-5324-11e8-94ef-d7c6ec3e04cd.png)

![selection_235](https://user-images.githubusercontent.com/26824900/39785271-314604dc-5324-11e8-8c80-4984abe396f6.png)

@droidmonkey @glubsy I've added "do not include" text input to gui and cli.
Buttons have precedence over this input.  If all characters form some group are mentioned in "do not include" field, then all characters from this group may appear in generated password. It is done like that to ensure that there's no inconsistency between selected groups and generated password.

glubsy commented:
@seregaxvm this is looking great in my opinion! Thank you!

droidmonkey commented:
Those are for hex passwords.

I don't understand the purpose of overriding the "do not include" entries if it covers the whole spectrum of a button. My opinion is if you put it in the do not include box, you mean it. 

I would add a clear button similar to what we have in the search box to quickly remove the do not include entries.

If you hide the advanced fields, does it revert back to normal behavior?

seregaxvm commented:
![selection_236](https://user-images.githubusercontent.com/26824900/39834493-ce79d268-53d5-11e8-9cac-6d91aca0e1f9.png)

![selection_237](https://user-images.githubusercontent.com/26824900/39834506-d97700f0-53d5-11e8-8e18-d7e3f1a1e8e3.png)

@droidmonkey  @glubsy
- Hexadecimal character buttons removed. One can generate hexadecimal passwords using "do not include" field. This is not very convenient but, as I understand, this feature is not so frequently used anyway.
- Clear button added to "do not include" field.
- "do not include" entries only applied when the field is visible. They are ignored in simple mode.
- "do not include" has precedence over buttons. Characters from this field are never in the password.
- When all characters, selected with buttons, are in "do not include" field, input is invalidated and no password will be generated. 
- Added `--exclude_similar` and `--every_group` options to CLI generator.
- Conversion `toLatin1` is used to calculate entropy so Unicode characters were not added.

glubsy commented:
Well this is looking very good. 

Perhaps the only thing that might confuse the user is when clicking `Advanced`, the `a-z` button is shifting below, and the `0-9` button is shifting left.

Also it might be a good opportunity to add examples in the `Exclude look-alike characters`. I, for one, don't know which characters are targeted by this.

droidmonkey commented:
Awesome update!

seregaxvm commented:
![selection_039](https://user-images.githubusercontent.com/26824900/39860332-95f37f0a-5445-11e8-83e2-c04d4a78a91a.png)

@glubsy added tool-tip description with list of excluded look-alike characters.
The last one is from Extended ASCII group. Should "." from special group also be excluded?
`a-z`, `A-Z` and `0-9` are grouped like so, firstly, to move related buttons closer to each other, secondly, to have columns of buttons with roughly same width.

glubsy commented:
Fantastic! Thank you very much!

droidmonkey commented:
I tried this out and it worked perfectly. Recommend the following additions/changes:

![adv_pwgen](https://user-images.githubusercontent.com/2809491/39970699-0d545388-56be-11e8-9aab-ea3f3836a09c.png)

The HEX button (oft requested feature) would very simply set the "exclude characters" textbox to ```GHIJKLMNOPQRSTUVWXYZghijklmnopqrstuvqrxyz``` it is up to the user to select the A-Z (or a-z) and 0-9 buttons.

seregaxvm commented:
@droidmonkey, please create branch `feature/finetune-password` so I could rebase to it?

seregaxvm commented:
![keepassxc_238](https://user-images.githubusercontent.com/26824900/39971713-84303c34-5708-11e8-98e8-8e8a1fc21ff7.png)
![keepassxc_239](https://user-images.githubusercontent.com/26824900/39971714-84541816-5708-11e8-80e5-b585da150317.png)

Checkboxes are hidden in simple mode. Added `Hex` shortcut. When clicked, it fills "exclude characters" textbox with predefined string `GHIJKLMNOPQRSTUVWXYZghijklmnopqrstuvwxyz`

droidmonkey commented:
@seregaxvm you create the branch on your own fork.

seregaxvm commented:
@droidmonkey  I did but I'm not able to change the brunch to pull from. I only can change base brunch to commit into, but it should exist.

droidmonkey commented:
Crap true, you cannot change the base branch. That is fine, I can merge it right now without much issue. In the future make sure you make  PR using a branch off develop.

droidmonkey commented on tests/TestCryptoHash.cpp:
```diff
@@ -50,8 +50,9 @@ void TestCryptoHash::test()
     QByteArray result3 = CryptoHash::hash(source2, CryptoHash::Sha512);
-    QCOMPARE(result3, QByteArray::fromHex("0d41b612584ed39ff72944c29494573e40f4bb95283455fae2e0be1e3565aa9f48057d59e6ff"
-                                          "d777970e282871c25a549a2763e5b724794f312c97021c42f91d"));
+    QCOMPARE(result3,
```
Revert these changes, your code formatter got a little happy.

droidmonkey commented on tests/TestCsvExporter.cpp:
```diff
@@ -64,9 +64,10 @@ void TestCsvExporter::testExport()
-    QString expectedResult =
-        QString().append(ExpectedHeaderLine).append("\"Test Group Name\",\"Test Entry Title\",\"Test Username\",\"Test "
-                                                    "Password\",\"http://test.url\",\"Test Notes\"\n");
+    QString expectedResult = QString()
```
Revert these changes, your code formatter got a little happy.

droidmonkey commented on tests/TestKdbx4.cpp:
```diff
@@ -317,11 +317,11 @@ void TestKdbx4::testUpgradeMasterKeyIntegrity_data()
     QTest::newRow("Upgrade (implicit): public-customdata") << QString("public-customdata") << KeePass2::FILE_VERSION_4;
-    QTest::newRow("Upgrade (implicit): rootgroup-customdata") << QString("rootgroup-customdata")
-                                                              << KeePass2::FILE_VERSION_4;
+    QTest::newRow("Upgrade (implicit): rootgroup-customdata")
```
Revert these changes, your code formatter got a little happy.

droidmonkey commented on tests/TestKdbx4.cpp:
```diff
@@ -317,11 +317,11 @@ void TestKdbx4::testUpgradeMasterKeyIntegrity_data()
     QTest::newRow("Upgrade (implicit): group-customdata") << QString("group-customdata") << KeePass2::FILE_VERSION_4;
-    QTest::newRow("Upgrade (implicit): rootentry-customdata") << QString("rootentry-customdata")
-                                                              << KeePass2::FILE_VERSION_4;
+    QTest::newRow("Upgrade (implicit): rootentry-customdata")
```
Revert these changes, your code formatter got a little happy.

seregaxvm commented:
@droidmonkey 
> Revert these changes, your code formatter got a little happy.

Done
> In the future make sure you make PR using a branch off develop.

Ok, sorry about that.

phoerious commented:
I'm confused. You can change the base branch at the top of this page.

seregaxvm commented:
Yes, I can change it, but only to branch that already exists. I have no write access, so I cannot create `feature/finetune-password` branch. If you would create it, I will be able to change base branch of this pr to `feature/finetune-password`.

droidmonkey commented:
I can change the target branch (ie develop) but not the base, GitHub doesnt even give me the option to try.

seregaxvm commented:
![selection_065](https://user-images.githubusercontent.com/26824900/40002282-ea412e08-5798-11e8-849a-aa5bd11cbbe4.png)

These are all my options. If it's possible I would prefer not to complicate things and to merge into `develop`. If it's not desirable, I will open another pull request.

TheZ3ro commented:
I really like this PR and the new UI (my 2 cent)

droidmonkey commented:
This is ready for merge. Any more changes requested? I'm very happy with the results as well.

TheZ3ro commented on src/core/PasswordGenerator.cpp:
```diff
@@ -220,6 +285,29 @@ QVector<PasswordGroup> PasswordGenerator::passwordGroups() const
     }
 
+    int i = 0;
+    while (1) {
```
It's not very clear at first look what this piece of code is suppose to do, insert a comment or move it into a function

seregaxvm replied:
How about that :"Loop over character groups and remove excluded characters from them; remove empty groups"? Also break condition should be moved from `if` to `while`

droidmonkey replied:
Also, you should use while(true) instead. You are implicitly converting 1 to bool in this case.

TheZ3ro replied:
@seregaxvm I think that can do it. :wink:

seregaxvm replied:
done

droidmonkey commented:
I will do a formal review very soon

Gittyperson commented:
It's great that these features are coming. Many, many sites accept the password during registration (containing the more unusual, non-allowed characters) but when it's time to log in, they will report "password is wrong" with no further details.

lordloh commented:
I would like to suggest adding a check box called "SQL safe" that would remove things like "--" and ";" from the password. I am not sure how developers handle them. I have had trouble logging into websites after using a password generated by Keepassxc. The password is accepted at change password time, but not during login :-( 

I had to tweak out certain sets of characters to get it working. This is completely the websites fault. But it would be ages before they fix it. I discovered at one site that the desktop site accepts longer passwords than the mobile version of the site. I had to use a shorter password :-(

droidmonkey commented:
Omg, if you are running into websites that can't handle those characters that means they are ripe for SQL injection attacks. RUN!

lordloh commented:
I am inclined to think that different sections section of their sites are
handling SQL injections differently.

Perhaps the order of sanitization and hashing passwords are different(?)

On Sat, Aug 11, 2018, 18:41 Jonathan White <notifications@github.com> wrote:

> Omg, if you are running into websites that can't handle those characters
> that means they are ripe for SQL injection attacks. RUN!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/keepassxreboot/keepassxc/pull/1841#issuecomment-412308474>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ACUnaxgJIIGUZ3jb-c-XSCJLQtCzBQD8ks5uP2uRgaJpZM4TSWU3>
> .
>

TheZ3ro commented:
> You can just type `;-)` or something like that into the `do not include` line.

The wink emoji of SQLi doom

sebast889 commented:
@seregaxvm @droidmonkey Sorry I couldn't find anywhere the reason so I'm asking here. Why are the character types in the group that they are in? Seems they are arbitrarily chosen or at least I couldn't spot any real pattern. Is there any rationale behind what characters are in what group?

droidmonkey commented:
This is already merged. If you have a suggested grouping(s) please out them in a new issue.

seregaxvm commented:
@sebast889 They are based on the personal experience with different sites password restrictions. The idea was to split `Special Characters` group into smaller ones. Their names are `Braces`, `Punctuation`, `Quotes`, `Math`, `Dashes` and `Logograms`.
